src/components/forms/components/DragAndDropUpload.js

```diff
@@ -0,0 +1,674 @@
+import React from 'react';
+import { Modal } from 'react-bootstrap';
+import PropTypes from 'prop-types';
+import { ajax } from './../../util';
+import _ from 'underscore';
+
+export class DragAndDropUploadSubmissionViewController extends React.Component {
+    /* Will become a submission-view specific version of the standalone controller to wrap
+    it and pass through unique versions of onUploadStart, patchToParent, etc f(x)s in DragAndDropFileUploadController */
+}
+
+/**
+ * A class utility for executing promise-chains in a sequential manner. Includes some
+ * scaffolding for aborting promises; needs more work in future.
+ *
+ * In drag-and-drop upload, this class ensures that each Item is uploaded and linked
+ * before starting the upload of the next item, so that the new atIds can be collected
+ * and patched to the parent together.
+ */
+class PromiseQueue {
```

Comment:
All the properties and methods of this class are static, meaning there is ever one single value/method shared between all instances of the class (technically is case with instance methods, given prototypical inheritance in JS objects/classes, but for instance methods, those methods are bound to `this` of the outer class/function scope.)

This is ok if want the class to be a singleton and only ever have one instance of it.. but I think we actually _might_ theoretically have multiple drag-drop areas or similar, so think would be better to have these as instance methods/properties (?).

Reply:
Hmm this is a good point. I was thinking, I think, that if someone started adding for Documents and then, while those things were being uploaded also started uploading some Images that they might be able to all end up in the same queue. I'm honestly not sure if that's how it would work, though, as I haven't been able to upload something big enough/that takes long enough to test it.

Reply:
Ok updated this to use instance-specific methods. And I create a new instance of it when it is used inside of the `DragAndDropFileUploadController.onUploadStart` method. So each separate Drag and Drop thing should now be using its own queue.